### ISSUE TITLE
Division by zero in floating point arithmetic.

### ISSUE THREAD

**Issue by asdasdasdasdasdf** [isocpp/CppCoreGuidelines#1524]
I propose that [ES.105: Don’t divide by zero](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es105-dont-divide-by-zero) should say "Don’t divide by zero in integer arithmetic"

This is to distinguish from division of floating point types for which division by zero is (I believe) well defined and sometimes useful. It seems like this is the intended meaning of the guideline based on its content though it's never explicitly said.

**Comment by jwakely** [comment]
The standard says it's undefined for all arithmetic and unscoped enumeration types. That includes floating point types.

**Comment by shaneasd** [comment]
Fascinating! Thanks for the reply.

Is this the bit of the standard you're talking about?
> If during the evaluation of an expression, the result is not mathematically defined or not in the range of representable values for its type, the behavior is undefined.
[ Note: Treatment of division by zero, forming a remainder using a zero divisor, and all floating-point exceptions vary among machines, and is sometimes adjustable by a library function.
— end note
 ]

Am I to understand then that the behavior is technically undefined but many compilers will give predictable results? In my experience msvc has always behaved in a "sensible" way and its [documentation](https://docs.microsoft.com/en-us/cpp/build/ieee-floating-point-representation?view=vs-2019) would seem to support this. Is this then in support of ```P.2: Write in ISO Standard C++``` i.e. I shouldn't rely on this behavior of msvc because it wont work on other compilers?

I think, regardless, that the rule could be clearer but I'm still trying to unravel the situation.

**Comment by jwakely** [comment]
No, I'm talking about [expr.mul] p4 which says "If the second operand or `/` or `%` is zero the behaviour is undefined."

**Comment by mbeutel** [comment]
Cf. relevant discussion at https://stackoverflow.com/questions/42926763/the-behaviour-of-floating-point-division-by-zero:

> Division by zero both integer and floating point are undefined behavior [[expr.mul]p4](http://eel.is/c++draft/expr.mul#4):
> 
> > The binary / operator yields the quotient, and the binary % operator yields the remainder from the division of the first expression by the second. If the second operand of / or % is zero the behavior is undefined. ...
> 
> Although implementation can optionally support [Annex F](http://port70.net/~nsz/c/c99/n1256.html#F) which has well defined semantics for floating point division by zero.
